File: C#/MethodInvocationRemotingMetricsTests/RemoteReceiverDecompressorMetricsTests.cs
/*
 * Copyright 2015 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#pragma warning disable 1591

using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using NMock2;
using ApplicationLogging;
using ApplicationMetrics;
using MethodInvocationRemoting;
using MethodInvocationRemotingMetrics;

namespace MethodInvocationRemotingMetricsTests
{
    //******************************************************************************
    //
    // Class: RemoteReceiverDecompressorMetricsTests
    //
    //******************************************************************************
    /// <summary>
    /// Unit tests for the metrics logging functionality in class MethodInvocationRemoting.RemoteReceiverDecompressor.
    /// </summary>
    [TestFixture]
    class RemoteReceiverDecompressorMetricsTests
    {
        private Mockery mocks;
        private IRemoteReceiver mockUnderlyingRemoteReceiver;
        private IMetricLogger mockMetricLogger;
        private RemoteReceiverDecompressor testRemoteReceiverDecompressor;

        [SetUp]
        protected void SetUp()
        {
            mocks = new Mockery();
            mockUnderlyingRemoteReceiver = mocks.NewMock<IRemoteReceiver>();
            mockMetricLogger = mocks.NewMock<IMetricLogger>();
            testRemoteReceiverDecompressor = new RemoteReceiverDecompressor(mockUnderlyingRemoteReceiver, mockMetricLogger);
        }

        [Test]
        public void ReceiveMetricsTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockUnderlyingRemoteReceiver).Method("Receive").Will(Return.Value("H4sIAAAAAAAEAO29B2AcSZYlJi9tynt/SvVK1+B0oQiAYBMk2JBAEOzBiM3mkuwdaUcjKasqgcplVmVdZhZAzO2dvPfee++999577733ujudTif33/8/XGZkAWz2zkrayZ4hgKrIHz9+fB8/Ih7/Hu8WZXqZ101RLT/7aHe881GaL6fVrFhefPbRuj3fPvjo9zh6/EXezqvZ2fKymmYttTSfvMgW+ZH8+t2inR+X5dOszd5cr/Lj5mVW07ctgX5812v92H3u/X702Lx4VCzb/CKvH9+1n/B3R7t79+Szo8d3vfdiIGZZm78pFvlxXWfXPUCPT8t8kS/bXvPHd7vfmKYR2D2wOzs7u9v8vzc7O4/4f2P6zOJsQb0HzIf0bO/ubd/bfbN379H9h/S/MX3Uh3lLwjRtTRPb6+YXlu3hm7xpX/PXv/CC/poXTUr/y9KWPk/lvRTt8O3vcffRR4cff/y97//iX/LJZ7//9p2tb/3CbLE6/P1+99/tx3/P3/Vnfl80uxvCuy2Cs3xaLLLythMnrW83b6ZtB+j2g4d7ewe7n+7d393f+3T/3r0H9x/eu79/7+H9nXv37t9q9gYgvyfgW1KoKS6W+ezJddvnloPbwphU1W1JjKa3oq807IA7z8pG8dxMw+jbbb2OvHzLMTbzqm7PBnTJ9u6n9/ZvrU6m86y+JbnQ9FbkkoYdcF/chlTRNyFyX5tUZbW8GKLUg0/v355Qs2o9KW+tdbnxrYhlmvZmcfzpwwefPry3e2//4NO9e3u790637+04MdgktHGQEYhRgLckyHlZZW0f77PlebEsWiXSLcHckqza5S2oOoDcvfHu/u79h3uf7hIxb0XLOCAD58HeaQzMLYc+ME/vS0GPxW9Jx0AobkHNTUK0TZaA9P/ezr1PD+7vP3hw/+B2TLoJZh/kg69N5KhK+V1v+7avaW9J3FA534K6m7X5vb0Hn96KoBvBAMrXp+HNFuwmMlrDflsiep7AbUg47DiQi3k78g2D2N37+qQbcqD2dx7eP/h05+H+/U/H9x4ePDzYf0D/d67TTQRl7/O2xFTP+DaEjDvRz6s6X6TFqlkv0llVVnXaFG0KrEbptFo2+bTN23WdZrNiVTRT+NN5WbTj9OU8a/KyXJPDvUyX67LM0kV2scxG6bJaptN13dBX0vQni8tsQX9d4k/pZZRmZfGL1jl1Nb7VHMaxf3y6WLXXeFV/ft25HApl9nZ2723v3Ed4tLv3aJ+imU/H5FvcFq5GhbecTRtD3mI6h+LN7b3d/Qf7B/c+3b+VbAyBsVA+WD5uOfbhoOiW3f1/RW6+ai3r/6J1trCvkfQ0DQWuxXJazNZL+jDF5NQz6mC1rlnO5M9xelJnzYeIzddG/sV6OU0vizbLVZhbVgEi8xbddb3MxulPrIuGhkkN58V0TeohW1/Qn/maRrqui40DuC2z+R+73ylF8wr4L92QB6X8rt/y7hfdjNH/A4VP6g5qEgAA"));
                Expect.Once.On(mockMetricLogger).Method("Begin").With(IsMetric.Equal(new StringDecompressTime()));
                Expect.Exactly(5).On(mockMetricLogger).Method("Increment").With(IsMetric.Equal(new RemoteReceiverDecompressorReadBufferCreated()));
                Expect.Once.On(mockMetricLogger).Method("End").With(IsMetric.Equal(new StringDecompressTime()));
                Expect.Once.On(mockMetricLogger).Method("Increment").With(IsMetric.Equal(new StringDecompressed()));
                Expect.Once.On(mockUnderlyingRemoteReceiver).Method("Receive").Will(Return.Value(""));
                Expect.Once.On(mockMetricLogger).Method("Begin").With(IsMetric.Equal(new StringDecompressTime()));
                Expect.Once.On(mockMetricLogger).Method("Increment").With(IsMetric.Equal(new RemoteReceiverDecompressorReadBufferCreated()));
                Expect.Once.On(mockMetricLogger).Method("End").With(IsMetric.Equal(new StringDecompressTime()));
                Expect.Once.On(mockMetricLogger).Method("Increment").With(IsMetric.Equal(new StringDecompressed()));
            }

            testRemoteReceiverDecompressor.Receive();
            testRemoteReceiverDecompressor.Receive();

            mocks.VerifyAllExpectationsHaveBeenMet();
        }

        [Test]
        public void ReceiveExceptionMetricsTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockUnderlyingRemoteReceiver).Method("Receive").Will(Return.Value("InvalidCompressedString"));
                Expect.Once.On(mockMetricLogger).Method("Begin").With(IsMetric.Equal(new StringDecompressTime()));
                Expect.Once.On(mockMetricLogger).Method("CancelBegin").With(IsMetric.Equal(new StringDecompressTime()));
            }

            Exception e = Assert.Throws<Exception>(delegate
            {
                testRemoteReceiverDecompressor.Receive();
            });
        }
    }
}

File: Java/MethodInvocationRemotingMetricsTests/src/net/alastairwyse/methodinvocationremotingmetricstests/RemoteReceiverDecompressorMetricsTests.java
/*
 * Copyright 2015 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.alastairwyse.methodinvocationremotingmetricstests;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.fail;
import static org.mockito.Matchers.isA;
import static org.mockito.Mockito.*;
import net.alastairwyse.applicationmetrics.*;
import net.alastairwyse.methodinvocationremoting.*;
import net.alastairwyse.methodinvocationremotingmetrics.*;

/**
 * Unit tests for the metrics logging functionality in class methodinvocationremoting.RemoteReceiverDecompressor.
 * @author Alastair Wyse
 */
public class RemoteReceiverDecompressorMetricsTests {

    private IRemoteReceiver mockUnderlyingRemoteReceiver;
    private IMetricLogger mockMetricLogger;
    private RemoteReceiverDecompressor testRemoteReceiverDecompressor;
    
    @Before
    public void setUp() throws Exception {
        mockUnderlyingRemoteReceiver = mock(IRemoteReceiver.class);
        mockMetricLogger = mock(IMetricLogger.class);
        testRemoteReceiverDecompressor = new RemoteReceiverDecompressor(mockUnderlyingRemoteReceiver, mockMetricLogger);
    }
    
    @Test
    public void ReceiveMetricsTest() throws Exception {
        when(mockUnderlyingRemoteReceiver.Receive())
            .thenReturn("H4sIAAAAAAAAAN1Y23LbNhD9FURJnTQ1bwCvNi3XTfzgmSSTtkr70NsgJCxhhgQVENBEk+TfA4qSIpukspLfqoFGBHBwAJzFLhdKLz+WBVowWfNKXIw82x0hJrIq52J6MdLq1opHl+P0NVOzKr8RiyqjyiA3LW9oycbt459cza6K4iVVdLKcs6v6LZWmVxnq1NlBp9/ad57H6WbgmAvFpkymzrZl1Tf2MGnbxqmzM66PIqeKTXjJrqSkyw5Rel2wkgnVgafO/Z4NtIe7Q+u6rmetysR1z1bFNm3bNW+pDuBMzMfysEW8CSZnQWKKbZq6nEBhaiWNYTvTnBTqfMJq9fuq+2RqajNeI1MoUqYdteNQg2t6L52z0fnTp3/98+nLTxf/WT8+e35Cy/n5vz88efzzo89/NzDnLh/YcizjJS2ghmvRMLttsPdIrSjBOPZCHHg+Dn1CoiAhgU+SwCUkAFlvgPlAYqgJ+VSw/Jel6p6WGMrxvqqgEjdQkL4t8B7dLS1qBtGwd7SSumcwVKdZJdXNQCyxvJD44HCSzagEytVAQXK1wHt0ryFS9Y5sXO5oqYpKTIeUisIALlRe6fcFOOquwDDfXUM7VrTDJAoT4hE/DjHBHrm2iBuDnLafsoexlxAoyG1RUdVd94245YKr5UE0QFnXUwJUHVgcsT3fCxIcekZMkJb9RBueCF/30Rx0pDrkhyq4c8SBOt5xCoCa+5zIMm8CE/+xS8I48KMoiGGHdB9nlzI6WuTekPLomEgLFPducAaouz+aExyFIEH30jQsx2v4/TcY+MUOFXEnE4BIOJw4mBQTJt8whYePl24ogfLdJIhDN/GD0CZJnMR+ZL4BWNBV9gkVc50ZQ4TsT6JfVZKViM9rXaK8KiqJaq5Qs6pTlFWiZpliSktEcz7nddbk06zgykZvZ7RmRaFNwi2Q0EVBUUmngp4iUQmUaVmbrhb6B1/Q0tQWTbWd5RTRgn/QzExlg2zYv/r0upyrZTN0/XusLYeuMtj1iOUGzfXIw2e+uc2EtsktoLz8oPjCDwgtQ/dNC3t+5Mck9EG+MUSzZXmwfzz4UvQ/85t3anv0P2habocZ76lrc3HlIuO5FqYRNcaRuZlgruXKz9qqjV5IWj/EbY5e/BstMrTgirK1M6tVCGh9frtcLQW10a+a12abBjjjmTbhgeqpqTJtdqol37sB6GFzep/rcfpbs37xbcuDXu7sIp3OP0ZfAYVP6g5qEgAA")
            .thenReturn("H4sIAAAAAAAAAAMAAAAAAAAAAAA=");

        testRemoteReceiverDecompressor.Receive();
        testRemoteReceiverDecompressor.Receive();

        verify(mockMetricLogger, times(2)).Begin(isA(StringDecompressTime.class));
        verify(mockMetricLogger, times(6)).Increment(isA(RemoteReceiverDecompressorReadBufferCreated.class));
        verify(mockMetricLogger, times(2)).End(isA(StringDecompressTime.class));
        verify(mockMetricLogger, times(2)).Increment(isA(StringDecompressed.class));
        verifyNoMoreInteractions(mockMetricLogger);
    }
    
    @Test
    public void ReceiveExceptionMetricsTest() throws Exception {
        when(mockUnderlyingRemoteReceiver.Receive()).thenReturn("InvalidCompressedString");
        
        try {
            testRemoteReceiverDecompressor.Receive();
            fail("Exception was not thrown.");
        }
        catch (Exception e) {
            verify(mockMetricLogger).Begin(isA(StringDecompressTime.class));
            verify(mockMetricLogger).CancelBegin(isA(StringDecompressTime.class));
            verifyNoMoreInteractions(mockMetricLogger);
        }
    }
}
